src/crypto/hpke/mod.rs

```diff
@@ -0,0 +1,236 @@
+//! Provides decryption primitives for HPKE according to the [`specification`].
+//!
+//! [`specification`]: https://github.com/patcg-individual-drafts/ipa/pull/31
+
+use hpke::aead::AeadTag;
+use hpke::{single_shot_open_in_place_detached, Deserializable, OpModeR};
+use std::{fmt, io};
+
+mod aad;
+mod registry;
+
+pub use aad::AssociatedData;
+pub use registry::KeyRegistry;
+
+/// IPA ciphersuite
+type IpaKem = hpke::kem::X25519HkdfSha256;
+type IpaAead = hpke::aead::AesGcm128;
+type IpaKdf = hpke::kdf::HkdfSha256;
+
+pub type KeyIdentifier = u8;
+pub type Epoch = u16;
+type IpaPublicKey = <IpaKem as hpke::kem::Kem>::PublicKey;
+type IpaPrivateKey = <IpaKem as hpke::kem::Kem>::PrivateKey;
+
+#[derive(Debug)]
+pub struct BottomError;
+
+impl From<hpke::HpkeError> for BottomError {
+    fn from(_value: hpke::HpkeError) -> Self {
+        Self
+    }
+}
+
+impl From<io::Error> for BottomError {
+    fn from(_value: io::Error) -> Self {
+        Self
+    }
+}
+
+impl fmt::Display for BottomError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Failed to open ciphertext")
+    }
+}
+
+/// Opens ciphertext in place by first obtaining the secret key from [`key_registry`]
+/// using epoch and key from [`aad`] and then applying [`HPKE decryption`] to the provided ciphertext.
+///
+/// ## Errors
+/// If ciphertext cannot be opened for any reason.
+///
+/// [`HPKE decryption`]: https://datatracker.ietf.org/doc/html/rfc9180#name-encryption-and-decryption
+pub fn open_in_place(
+    key_registry: &KeyRegistry,
+    encap_key: &[u8],
+    ciphertext: &mut [u8],
+    tag: &[u8],
```

Comment:
I know that the in-place function needs to have two slices for the ciphertext and tag, but my preference is that we do the split for them.  That is, we take `ciphertext: &mut [u8]` here and do:

```rust
let (ct, tag) = ciphertext.split_at_mut(ciphertext.len() - IpaAead::N_T);
```
(Though sadly the constants defined in the spec aren't exposed on the AEAD trait in the crate.  Maybe we can upstream something...; until then, a constant in this code (TAG_LEN) should do.)

Then you can return an immutable reference to `ct`.  Note that you need to use lifetime annotations so that the compiler gets it right.

Reply:
synced up offline, we are keeping the match key encryption as [enc, ct] where ct = ct + tag. Once we start consuming it from the body, we will be able to see whether it is more convenient to keep it as is or to include the encapped key into the same slice